File: frontend/src/types/metadataSources.ts
export type MetadataSourceCategory = 'local' | 'installed_tool' | 'external_api' | 'external_input'
export type MetadataConnectionStatus = 'not_tested' | 'unavailable' | 'ready' | 'failed' | 'not_implemented'

export interface MetadataSource {
  id: string; label: string; category: MetadataSourceCategory; enabled: boolean; configured: boolean
  requires_credentials: boolean; credentials_status: 'not_required' | 'missing' | 'saved' | 'invalid' | 'unknown'
  credential_fields: string[]; saved_credential_fields: string[]; connection_status: MetadataConnectionStatus
  priority: number; best_for: string[]; current_behavior: 'implemented' | 'preview_only' | 'settings_only' | 'planned'
  configuration_note?: string | null; safety: string[]
}
export interface MetadataSourcesResponse { sources: MetadataSource[] }
export interface MetadataSourceUpdate { id: string; enabled?: boolean; priority?: number; credentials?: Record<string, string | null> }
export interface MetadataSourceTestResult { source_id: string; connection_status: MetadataConnectionStatus; message: string; network_used: boolean }
export interface MetadataSourceClearResult { source_id: string; cleared: boolean; message: string }

File: backend/app/schemas/metadata_sources.py
"""Safe request/response shapes for local metadata-source settings."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class MetadataSourceUpdate(BaseModel):
    id: str = Field(min_length=1, max_length=64)
    enabled: bool | None = None
    priority: int | None = Field(default=None, ge=1, le=1000)
    credentials: dict[str, str | None] | None = None


class MetadataSourcesUpdateRequest(BaseModel):
    sources: list[MetadataSourceUpdate] = Field(min_length=1, max_length=10)


class MetadataSourceSummary(BaseModel):
    """Safe, server-owned metadata-source capability summary."""

    id: str
    label: str
    category: Literal["local", "installed_tool", "external_api", "external_input"]
    role: Literal["local_input", "analysis_only", "track_enrichment"]
    enabled: bool
    configured: bool
    requires_credentials: bool
    credentials_status: Literal["not_required", "missing", "saved", "invalid", "unknown"]
    credential_fields: list[str]
    saved_credential_fields: list[str]
    connection_status: Literal["not_tested", "unavailable", "ready", "failed", "not_implemented", "needs_setup", "misconfigured"]
    needs_setup: bool
    selectable_for_enrichment: bool
    priority: int
    best_for: list[str]
    current_behavior: Literal["implemented", "preview_only", "settings_only", "planned"]
    configuration_note: str | None = None
    safety: list[str]


class MetadataSourcesResponse(BaseModel):
    sources: list[MetadataSourceSummary]


class MetadataSourceTestResponse(BaseModel):
    source_id: str
    connection_status: Literal["not_tested", "unavailable", "ready", "failed", "not_implemented"]
    message: str
    network_used: bool = False


class MetadataSourceClearResponse(BaseModel):
    source_id: str
    cleared: bool
    message: str
